esmvalcore/config/_config_object.py

```diff
@@ -233,23 +252,90 @@ def _get_config_path_from_cli() -> None | str:
 
         return None
 
+    # TODO: remove in v2.14.0
     def load_from_file(
         self,
         filename: Optional[os.PathLike | str] = None,
     ) -> None:
-        """Load user configuration from the given file."""
+        """Load user configuration from the given file.
+
+        .. deprecated:: 2.12.0
+            This method has been deprecated in ESMValCore version 2.14.0 and is
+            scheduled for removal in version 2.14.0. Please update
+            `esmvalcore.config.CFG` directly instead using `CFG.update()` or
+            `CFG[...] = ...`.
+
+        Parameters
+        ----------
+        filename:
+            YAML file to load.
+
+        """
+        msg = (
+            "The method `CFG.load_from_file()` has been deprecated in "
+            "ESMValCore version 2.12.0 and is scheduled for removal in "
+            "version 2.14.0. Please update `esmvalcore.config.CFG` directly "
+            "instead using `CFG.update()` or `CFG[...] = ...`."
+        )
+        warnings.warn(msg, ESMValCoreDeprecationWarning)
         self.clear()
         self.update(Config._load_user_config(filename))
 
     def reload(self):
-        """Reload the config file."""
-        if 'config_file' not in self:
-            raise ValueError(
-                "Cannot reload configuration, option 'config_file' is "
-                "missing; make sure to only use the `CFG` object from the "
-                "`esmvalcore.config` module"
+        """Reload the configuration object.
+
+        This considers the following YAML files (descending priority):
+
+        1. If set, the directory specified with the ``ESMVALTOOL_CONFIG_DIR``
+           environment variable.
+
+        2. The user configuration directory (by default
+           ``~/.config/esmvaltool``, but this can be changed with the
+           ``--config_dir`` command line argument when running the
+           ``esmvaltool`` command line tool).
+
+        3. Default configuration, see :ref:`config_options`.
```

Comment:
Would it be possible to change this to:

        1. The ``--config_dir`` command line argument when running the
           ``esmvaltool`` command line tool

        2. The user configuration directory: by default
           ``~/.config/esmvaltool``, but this can be changed with the
           ``ESMVALTOOL_CONFIG_DIR` environment variable.
        3. Default configuration, see :ref:`config_options`.

Reply:
Yes, this is certainly possible and probably also more straightforward. I opted for the current version to stick as closely as possible to the old way of specifying the configuration (where the `--config-file` argument overwrote the default user config location `~/.esmvaltool/config-user-yml`).

Reply:
Done in https://github.com/ESMValGroup/ESMValCore/pull/2448/commits/2decd64f0b43f58cb8145c1fd4b1cbd42decf9a6.

Turns out that changing this allowed me to get rid of all the complicated code that took care of parsing the CLI arguments (>200 lines!): if the environment variable `ESMVALTOOL_CONFIG_DIR` can be used to replace the default user dir instead of the CLI argument, no special parsing is needed.

Right now the tool will NOT raise an error if the user config dir (the one in home or given by the env variable) does not exist (there will be a message at the very beginning that the dir does not exist). This was much easier to implement this way and the behavior is identical to when the ESMValTool Python API is used. When a `--config_dir` is specified and it does not exist, the tool WILL raise an error.